### PR TITLE
Add PROVIDED DependencyMode

### PR DESCRIPTION
**Pull request by xa17d**
First step into the direction of retiring `MOCK`, `REAL` dependency modes and differentiation between `offer` and `require`.

### CODE REVIEW

File: sweetest/src/main/java/com/mysugr/sweetest/framework/build/BaseBuilder.kt [hunk 1]
```diff
@@ -83,21 +144,39 @@ abstract class BaseBuilder<TSelf, TResult : BaseAccessor>(
         testContext.factories.configure(FactoryRunner0(R::class.java, createObject))
     }
 
-    inline fun <reified T : Steps, reified R : Any> offerFactory(noinline createObject: (T) -> R) = apply {
-        testContext.factories.configure(FactoryRunner1(R::class.java, T::class.java, createObject))
-    }
+    inline fun <reified T : Steps, reified R : Any> offerFactory(noinline createObject: (T) -> R) =
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/build/BaseBuilder.kt:147]: i think auto format changed the line breaks. can you change the code style settings in your IDE to 120 characters and revert these changes?

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 2]
```diff
@@ -7,16 +7,19 @@ import kotlin.reflect.KClass
 
 class DependenciesTestContext {
 
+    @Deprecated("Use \"provide\" instead.")
     fun requireReal(clazz: KClass<*>) {
         val dependency = getDependencyConfiguration(clazz)
         TestEnvironment.dependencies.states[dependency].mode = DependencyMode.REAL
     }
 
+    @Deprecated("Use \"provide\" instead.")
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:16]: you could add a `ReplaceWith` here...

File: sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt [hunk 3]
```diff
@@ -43,11 +49,25 @@ class DependenciesTestContext {
...
     fun requireSpy(clazz: KClass<*>) {
         val dependency = getDependencyConfiguration(clazz)
         TestEnvironment.dependencies.states[dependency].mode = DependencyMode.SPY
     }
 
     private fun getDependencyConfiguration(clazz: KClass<*>) =
         TestEnvironment.dependencies.configurations.getAssignableTo(clazz)
+
+    fun provide(clazz: KClass<*>, initializer: DependencyInitializer<*>) {
```
Comment (eder7) [inline sweetest/src/main/java/com/mysugr/sweetest/framework/context/DependenciesTestContext.kt:61]: maybe you can add some kdoc here

File: sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+import com.mysugr.sweetest.framework.configuration.moduleTestingConfiguration
+import com.mysugr.sweetest.framework.context.TestContext
+import com.mysugr.sweetest.framework.environment.TestEnvironment
+import com.mysugr.sweetest.util.mock
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+import org.mockito.Mockito.`when`
+
+class ProvideTest {
```
Comment (eder7) [inline sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt:17]: for simplicity i thing you could have merged steps and tests together into one, because under the hood they use the same accessor anyway, but no strong opinion. feel free to have a look at [my test](https://github.com/mysugr/sweetest/pull/28/files#diff-ed85968250e216f3c9fb5e3bd4210dc2) though...

File: sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt [hunk 5]
```diff
@@ -35,155 +34,128 @@ class ProvideTest {
...
+            val viewModel by dependency<TestViewModel>()
         }
-    }
 
-    private fun givenServiceConfiguredAsMockOnly() {
-        moduleTestingConfiguration = moduleTestingConfiguration {
-            dependency mockOnly of<TestUserService>()
-            dependency any of<TestViewModel>()
+        TestClass().run {
+            junitBefore()
```
Comment (eder7) [inline sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt:154]: hm, can you check at which point the exception is thrown? i think it should fail happen here. but then the subsequent lines can be removed...

Comment (eder7) [reply sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt:154]: i tried it and in fact the test still works when you remove the lines below.

File: sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+            assertEquals(TestViewModel::class.java, viewModel::class.java)
+            assertEquals(FakeTestUserService::class.java, userService::class.java)
+            assertEquals(userService, viewModel.userService)
+            assertFalse(viewModel.isMock)
+        }
+    }
+
+    @Test
+    fun `provide works together with mockOnly and requireMock`() { // TODO MARKER
```
Comment (ChristopherKoellner) [inline sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt:84]: There is still a TODO marker here. 
IS there a reason for it?

File: sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt (hunk 5 above)
Comment (eder7) [reply sweetest/src/test/java/com/mysugr/sweetest/ProvideTest.kt:136]: did it on my own
